### PR TITLE
Integrate locations

### PR DESCRIPTION
**Pull request by ikelesoglu**
- Add LocationData class into PokedexScreenState
- Bind Locations with retreiveLocations

### CODE REVIEW

File: domain/src/main/java/com/pokedroid/domain/model/Location.kt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package com.pokedroid.domain.model
+
+class Location(val lat: Int, val lon: Int) {
```
Comment (fabiocarballo) [inline domain/src/main/java/com/pokedroid/domain/model/Location.kt]: I should have give you a bit more context!

https://pokeapi.co/docsv2/#locations-section

Check out this part of the API and try to figure out some interesting info that we could use.

File: presentation/src/main/java/com/pokedroid/presentation/PokedexScreenState.kt [hunk 2]
```diff
@@ -3,6 +3,7 @@ package com.pokedroid.presentation
 sealed class PokedexScreenState {
 
     data class Data(val numPokemons: Int): PokedexScreenState()
+    data class LocationData(val numLocations: Int): PokedexScreenState()
```
Comment (fabiocarballo) [inline presentation/src/main/java/com/pokedroid/presentation/PokedexScreenState.kt]: Unless it is really hard to do (or feel really bad) we should try to just have one `Data` class! With that said I would suggest for you to move the `numLocations` into the `Data` one.

File: presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt [hunk 3]
```diff
@@ -2,31 +2,52 @@ package com.pokedroid.presentation
...
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
+import com.pokedroid.domain.interactors.RetrieveLocations
 import com.pokedroid.domain.interactors.RetrievePokemons
 import io.reactivex.Observable
 import io.reactivex.disposables.CompositeDisposable
 import io.reactivex.disposables.Disposable
 
-class PokedexViewModel(private val retrievePokemons: RetrievePokemons): ViewModel() {
+/**
```
Comment (fabiocarballo) [inline presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt]: The comments can actually be deleted now!

File: presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt [hunk 4]
```diff
@@ -2,31 +2,52 @@ package com.pokedroid.presentation
...
 
     private fun bindPokemons(): Disposable {
         val pokedexObservable: Observable<PokedexScreenState> =
-            retrievePokemons
-                .retrieveBehaviorStream(Unit)
-                .map { PokedexScreenState.Data(it.size) }
+                retrievePokemons
+                        .retrieveBehaviorStream(Unit)
+                        .map {
+                            PokedexScreenState.Data(it.size)
```
Comment (fabiocarballo) [inline presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt]: when we have small expressions we prefer (in terms of style) to use:

`.map { PokedexScreenState.Data(it.size) }`

File: presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt [hunk 5]
```diff
@@ -2,31 +2,52 @@ package com.pokedroid.presentation
...
         return pokedexObservable
-            .startWith(PokedexScreenState.Loading)
-            .onErrorReturn { PokedexScreenState.Error("Ups!") }
-            .subscribe(pokedexLiveData::postValue, ::handleError)
+                .startWith(PokedexScreenState.Loading)
+                .onErrorReturn { PokedexScreenState.Error("Ups!") }
+                .subscribe(pokedexLiveData::postValue, ::handleError)
+    }
+
+    private fun bindLocations(): Disposable {
```
Comment (fabiocarballo) [inline presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt]: Well done. The thing here is that you are building two streams of data (one for list of pokemons, another for list of locations).

As a next step in this PR I want you to take a look on how you can combine two streams of data into one (check for instance how to use the [RxJava Zip](http://reactivex.io/documentation/operators/zip.html) operator) .

Also, a good start on RxJava for you could be this one: https://www.youtube.com/watch?v=YPf6AYDaYf8

**Comment by ikelesoglu** [comment]
- Only added id and name of the location for now, will check API in more detail soon.
- Used zip to combine two streams and idk why I used listOf :D I thought it might make sense to create a list of two lists basically but I'm not fully sure how will map work wit this list.

File: presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt [hunk 6]
```diff
@@ -20,36 +20,22 @@ class PokedexViewModel(private val retrievePokemons: RetrievePokemons, private v
...
+        val retrievedPokemons = retrievePokemons.retrieveBehaviorStream(Unit)
+        val retrievedLocations = retrieveLocations.retrieveBehaviorStream(Unit)
+
         val pokedexObservable: Observable<PokedexScreenState> =
-                retrievePokemons
-                        .retrieveBehaviorStream(Unit)
-                        .map {
-                            PokedexScreenState.Data(it.size)
-                        }
+                zip(retrievedPokemons, retrievedLocations, BiFunction { pokeList: PokemonList, locList: List<Location> -> listOf(pokeList, locList) })
```
Comment (fabiocarballo) [inline presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt]: Zipping like a PRO!

Comment (fabiocarballo) [reply presentation/src/main/java/com/pokedroid/presentation/PokedexViewModel.kt]: and meanwhile noticed we could do a small improvement here:

```
zip(retrievedPokemons, retrievedLocations, BiFunction { pokeList: PokemonList, locList: List<Location> ->
     PokedexScreenState.Data(pokeList.size, locList.size)
}
```

You can do the mapping right away in the transformation function in the zip!

**Pull request merged**